rbx_dom_weak/src/dom.rs

```diff
@@ -235,6 +235,48 @@ impl WeakDom {
         dest_parent.children.push(referent);
     }
 
+    /// Clone the instance with the given `referent` and all its descendants
+    /// (i.e. the entire subtree) into the same WeakDom.
+    ///
+    /// After the operation, the root of the cloned subtree has no parent.
+    ///
+    /// Any Ref properties that point to instances contained in the subtree are
+    /// rewritten to point to the cloned instances.
+    pub fn clone_within(&mut self, referent: Ref) -> Ref {
+        let mut ctx = CloneContext::default();
+        let root_builder = ctx.clone_ref_as_builder(self, referent);
+        let root_ref = self.insert(Ref::none(), root_builder);
+
+        while let Some((cloned_parent, uncloned_child)) = ctx.queue.pop_front() {
+            let builder = ctx.clone_ref_as_builder(self, uncloned_child);
+            self.insert(cloned_parent, builder);
+        }
+
+        ctx.rewrite_refs(self);
+        root_ref
+    }
+
+    /// Clone the instance with the given `referent` and all its descendants (i.e. the
+    /// entire subtree) into the given WeakDom.
+    ///
+    /// After the operation, the root of the cloned subtree has no parent.
+    ///
+    /// Any Ref properties that point to instances contained in the subtree are
+    /// rewritten to point to the cloned instances.
+    pub fn clone_into_external(&self, referent: Ref, dest: &mut WeakDom) -> Ref {
+        let mut ctx = CloneContext::default();
+        let root_builder = ctx.clone_ref_as_builder(self, referent);
+        let root_ref = dest.insert(Ref::none(), root_builder);
+
+        while let Some((cloned_parent, uncloned_child)) = ctx.queue.pop_front() {
+            let builder = ctx.clone_ref_as_builder(self, uncloned_child);
+            dest.insert(cloned_parent, builder);
+        }
+
+        ctx.rewrite_refs(dest);
```

Comment:
Cloning into an external DOM and leaving a bunch of Ref properties that point to nothing (despite not being `null`) makes me leery, so I went and checked what our serializers do when they run into one.

`rbx_binary` handles orphaned referents invisibly and correctly by just writing them equivalently to `null` but `rbx_xml` doesn't, and will increment the internal referent counter by 1 whenever it runs into a non-null referent. As a result, orphaned Ref properties will cause the rest of the file to be off by one. This isn't a problem since the deserializer skips unknown referents, but it makes this method for cloning not invisible to consumers like Lune.

We can either fix that behavior in `rbx_xml`'s serializer or rewrite orphaned Ref properties to be `None`, or do both. I'm inclined to say both but doing it in `rbx_xml` is outside of scope for this PR.